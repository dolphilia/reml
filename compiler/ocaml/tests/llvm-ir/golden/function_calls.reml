// Function calls and ABI test (LLVM IR golden test)
//
// Tests:
// - Multiple argument passing
// - Function return values
// - System V ABI compliance
// - Function composition

fn add(a, b) -> i64 = a + b

fn mul(a, b) -> i64 = a * b

fn add3(a, b, c) -> i64 = a + b + c

fn compute(x, y, z) -> i64 {
  let sum = add(x, y)
  let prod = mul(sum, z)
  prod
}

let result1 = add3(10, 20, 30)
let result2 = compute(5, 10, 3)
