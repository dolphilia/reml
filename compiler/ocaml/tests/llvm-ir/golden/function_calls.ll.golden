; ModuleID = 'main'
source_filename = "main"
target datalayout = "e-m:e-p:64:64-f64:64:64-v128:128:128-a:0:64"
target triple = "x86_64-unknown-linux-gnu"

declare ptr @mem_alloc(i64)

declare void @inc_ref(ptr)

declare void @dec_ref(ptr)

; Function Attrs: noreturn
declare void @panic(ptr, i64) #0

declare void @print_i64(i64)

define i64 @add(i64 %a, i64 %b) {
entry:
  %add_tmp = add i64 %a, %b
  ret i64 %add_tmp
}

define i64 @mul(i64 %a, i64 %b) {
entry:
  %mul_tmp = mul i64 %a, %b
  ret i64 %mul_tmp
}

define i64 @add3(i64 %a, i64 %b, i64 %c) {
entry:
  %add_tmp = add i64 %a, %b
  %add_tmp1 = add i64 %add_tmp, %c
  ret i64 %add_tmp1
}

define i64 @compute(i64 %x, i64 %y, i64 %z) {
entry:
  %call_tmp = call i64 @add(i64 %x, i64 %y)
  %call_tmp1 = call i64 @mul(i64 %call_tmp, i64 %z)
  ret i64 %call_tmp1
}

attributes #0 = { noreturn }
